src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java

```diff
@@ -0,0 +1,116 @@
+    private static JabRefItemDataProvider dataProvider = new JabRefItemDataProvider();
+    private static CSL cslInstance;
+
+    /**
+     * Creates the bibliography of the provided items. This method needs to run synchronized because the underlying
+     * CSL engine is not thread-safe.
+     */
+    static synchronized List<String> makeBibliography(List<BibEntry> bibEntries, String style, CitationStyleOutputFormat outputFormat) throws IOException {
+        initialize(style, outputFormat);
+        List<CSLItemData> cslItemData = new ArrayList<>(bibEntries.size());
+        for (BibEntry bibEntry : bibEntries) {
+            cslItemData.add(bibEntryToCSLItemData(bibEntry));
+        }
+        dataProvider.setData(cslItemData);
+        cslInstance.registerCitationItems(dataProvider.getIds());
+        final Bibliography bibliography = cslInstance.makeBibliography();
+        return Arrays.asList(bibliography.getEntries());
+    }
+
+    /**
+     * Initialized the static CSL instance if needed.
+     *
+     * @param newStyle  journal style of the output
+     * @param newFormat usually HTML or RTF.
+     * @throws IOException An error occurred in the underlying JavaScript framework
+     */
+    private static void initialize(String newStyle, CitationStyleOutputFormat newFormat) throws IOException {
+        if (cslInstance == null || !Objects.equals(newStyle, CSLAdapter.style) || !Objects.equals(newFormat, CSLAdapter.format)) {
+            cslInstance = new CSL(dataProvider, newStyle);
+            style = newStyle;
+        }
+
+        if (!Objects.equals(newFormat, CSLAdapter.format)) {
+            cslInstance.setOutputFormat(newFormat.getFormat());
+            format = newFormat;
+        }
+    }
+
+    /**
+     * Converts the {@link BibEntry} into {@link CSLItemData}.
+     */
+    private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry) {
+        String citeKey = bibEntry.getCiteKeyOptional().orElse("");
+        BibTeXEntry bibTeXEntry = new BibTeXEntry(new Key(bibEntry.getType()), new Key(citeKey));
+
+        // Not every field is already generated into latex free fields
+        for (String key : bibEntry.getFieldMap().keySet()) {
+            Optional<String> latexFreeField = bibEntry.getLatexFreeField(key);
+            latexFreeField.ifPresent(value -> bibTeXEntry.addField(new Key(key), new DigitStringValue(value)));
+        }
+        return BIBTEX_CONVERTER.toItemData(bibTeXEntry);
+    }
+
+    /**
+     * Custom ItemDataProvider that allows to set the data so that we don't have to instantiate a new CSL object
+     * every time.
+     */
+    private static class JabRefItemDataProvider implements ItemDataProvider {
+
+        private ItemDataProvider dataProvider = new ListItemDataProvider();
+
+        public void setData(List<CSLItemData> data) {
+            dataProvider = new ListItemDataProvider(data);
+        }
+
+        @Override
+        public CSLItemData retrieveItem(String id) {
```

Comment:
I'm not sure if this is really an advantage but since you are already implementing a custom data provider, I would suggest the following:
- setData accepts a list of `BibEntry` 
- retriveItem gets the `BibEntry` with the given key and then returns the result of `bibEntryToCSLItemData`. This solution might be a bit more memory effective since you don't have an additional list (but it only makes sense if the `bibEntryToCSLItemData` method is not expensive).

Reply:
Ha funny. It's the exact same thing I have thought about. I wanted to wait until we are sure it works in general. I too would prefer to make a real `JabRefItemDataProvider` that works on `BibEntry`.

Reply:
I refactored this. `retrieveItem` cannot change its signature because it is called internally by CSL (that's why it is overridden). Still, the layout is much clearer now.